### DISCUSSION TITLE
How to turn-off event sending after reconnection?

### DISCUSSION THREAD

**Discussion by hyperknot** [socketio/socket.io#3729] [Q&A] [answered]
I would like to send non-volatile (thus "normal") events, but I'd like to turn off the "remember events and send them after reconnection" behaviour. How is it possible to achieve this?

**Comment by hyperknot** [comment]
So far what seems to be working is to keep a this.connected variable and set it to false in a socket.on('disconnect') event. Then wrap every emit in a `if (this.connected) {socket.emit()}` block.

From my experimentation it seems that:
- events which are sent before "realizing" the disconnection fact (first unanswered ping) simply disappear
- events after the disconnection are not sent because of the client side logic

Can someone confirm this is the right logic to do this?

**Comment by darrachequesne** [comment, marked as answer]
There are several solutions, depending on your use case:

- use the [connected](https://socket.io/docs/v3/client-socket-instance/#Socket-connected) attribute (as you suggested)

```js
if (socket.connected) {
  socket.emit( /* ... */ );
} else {
  // ...
}
```

- empty the internal buffer upon connection

```js
socket.on("connect", () => {
  socket.sendBuffer = [];
});
```

- use the [volatile](https://socket.io/docs/v3/emitting-events/#Volatile-events) flag

```js
socket.volatile.emit( /* ... */ );
```

I think this is currently missing from the documentation, I will definitely add this.

**Reply by hyperknot** [reply]
Thanks a lot for clarifying this!

**Reply by darrachequesne** [reply]
You're welcome! I've added it here: https://socket.io/docs/v3/client-offline-behavior/
